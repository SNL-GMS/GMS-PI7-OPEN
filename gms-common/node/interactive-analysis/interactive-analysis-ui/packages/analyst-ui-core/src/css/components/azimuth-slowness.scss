@import "~@blueprintjs/core/lib/scss/variables.scss";
.fk-thumbnail {
    outline: 1px solid $gray3;
    position: relative;
    margin-left: 0.22rem;
    margin-right: 0.22rem;
    margin-bottom: 0.44rem;
    &:hover {
        outline: 1px solid $red2;
        cursor: pointer;
    }
    &.selected {
        outline: 3px solid $red3;
    }
    .fk-thumbnail__label {
        color: $black;
        background-color: $gold5;
        position: absolute;
        top: 0px;
        left: 0px;
        padding: 0.15rem;
        z-index: 1;
        font-size: 0.6rem;
    }
    .fk-thumbnail__label--reviewed {
        background-color: $dark-gray3;
        position: absolute;
        top: 0px;
        left: 0px;
        padding: 0.15rem;
        z-index: 1;
        font-size: 0.6rem;
    }
    &:hover .fk-thumbnail__label {
        background-color: $red3;
    }
}
.fk-thumbnail__canvas {
    position: absolute;
    left: 0px;
    right: 0px;
    top: 0px;
    bottom: 0px;
    z-index: 0
}
.azimuth-slowness-data-display {
    width: 100%;
    overflow-y: auto;
    padding: .5rem;
}
.azimuth-slowness-data-display__wrapper {
    flex: 0 0 auto;
    display: flex;
    margin-bottom: 8px;
}
.azimuth-slowness-divider {
    position: relative;
    width: 0.5rem;
    cursor: col-resize;
    flex: 0 0 auto;

}
.azimuth-slowness-divider:hover {
    div {
        background-color: $forest3;
    }
}
.azimuth-slowness-divider__spacer {
    position: absolute;
    left: 0.2rem;
    right: 0.2rem;
    top: 0px;
    bottom: 0px;
}

.fk-axis path,
.fk-axis line {
    fill: none;
    stroke: #ddd;
    shape-rendering: crispEdges;
}

.fk-axis text {
    font-family: sans-serif;
    font-size: 11px;
    fill: #ddd;
}

.fk-weavess-wrapper {
    position: absolute;
    top: 1px;
    bottom: 1px;
    left: 1px;
    right: 4px;
    min-height: 250px;
    color: $gray4;
}
.azimuth-slowness-container {
    padding: 0.5rem;
    height: 100%;
    width: 100%;
    overflow: hidden;
    display: flex;
    user-select: none;
}
.azimuth-slowness-thumbnails {
    flex: 0 0 auto;
    display: flex;
    flex-direction: column;
    overflow: hidden;
}
.azimuth-slowness-thumbnails__control-container {
    position: relative;
    margin-bottom: 0.5rem;
    height: auto;
}
.azimuth-slowness-thumbnails__wrapper-1 {
    flex: 1 1 auto;
    position: relative;
    height: auto;
    overflow: auto;
}
.azimuth-slowness-thumbnails__wrapper-2 {
    width: 100%;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    margin: auto;
}
.azimuth-slowness-thumbnails__wrapper-3 {
    padding: 0.25rem;
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
}
.azimuth-slowness-thumbnails-controls__wrapper {
    display: inline-flex;
    flex-direction: row;
    flex: 0 1 auto;
}
.azimuth-slowness-thumbnails-controls {
    display: inline-flex;
    flex-direction: row;
    flex: 0 1 auto;
    align-items: baseline;
}
.azimuth-slowness-thumbnails-controls__label {
    padding-left: 10px;
    padding-right: 6px;
}

.fk-properties {
    height: 100%;
    overflow: auto;
    min-width: 300px;
    max-width: 660px;
    flex: 2 1 auto;
}

.fk-properties__column {
    height: 100%;
    display: flex;
    flex-direction: column;
    overflow: hidden;
}
.fk-properties-label-row {
    min-height: 1.2em;
    display: flex;
    flex-direction: row;
    align-items: baseline;
    justify-content: space-between;
    margin-bottom: 4px;
}
.fk-properties-label-row__left {
    display: flex;
    justify-content: flex-start;
}
.fk-properties__label {
    padding-left: 4px;
    margin-right: 24px;
}
.fk-plots__toolbar {
    margin-bottom: 4px;
    margin-right: 2px;
}
.fk-properties__table {
    flex: 1 1 auto;
    position: relative;
    min-height: 120px;
    max-width: 100%;
}
.fk-controls {
    height: 100%;
    display: flex;
    flex-direction: row;
    padding-top: 8px;
}
.fk-control__grid {
    display: grid;
    grid-template-columns: auto auto;
    justify-content: space-between;
    grid-column-gap: 8px;
    grid-row-gap: 4px;
    align-items: baseline;
}
.fk-control__button {
    margin-top: 24px;
}
.fk {
    display: flex;
    flex: 0 0 auto;
    flex-direction: column;
    width: 300px;
    height: 300px;
    margin-right: 1rem;
}
.fk-rendering {
    position: relative;
    width: 100%;
    height: 100%;
    margin-right: 0.5rem;
    margin-bottom: 0.5rem;
}
.fk-rendering__y-axis {
    position: absolute;
    top: -10px;
    left: 0px;
    bottom: 25px;
    width: 35px;
}
.fk-rendering__x-axis {
    position: absolute;
    right: -10px;
    left: 25px;
    bottom: 0px;
    height: 35px;
}
.fk-rendering__slowness {
    position: absolute;
    left: 35px;
    right: 0px;
    bottom: 0px;
    font-size: 0.5rem;
    text-align: center;
    z-index: 1;
}
.fk-rendering__slowness-label {
    height: calc(100px + 0.5rem);
    width: 260px;
    padding: 0.25rem;
}
.fk-legend {
    height: 50px;
}
.fk-legend__canvas {
    position: absolute;
    right: 100px;
    left: 100px;
    top: 0px;
    z-index: 0;
}
.fk-legend-font__analyst {
    position: absolute;
    left: 120px;
    font-size: $pt-font-size-small;
}
.fk-legend-font__peak {
    position: absolute;
    left: 120px;
    top: 20px;
    font-size: $pt-font-size-small;
}
.fk-legend-font__predicted {
    position: absolute;
    left: 120px;
    top: 36px;
    font-size: $pt-font-size-small;
}
.fk-rendering__canvas {
    position: absolute;
    right: 0px;
    top: 0px;
    z-index: 0;
}
.fk-plot-controls {
    display: inline-flex;
    flex-direction: row;
    align-items: baseline;
    flex: 0 0 auto;
    padding-bottom: 0.25rem;
}
.fk-plot-window-select-control { 
    display: flex;
    flex-direction: row;
    align-items: baseline;
    flex: 0 1 auto;
    padding-right: 6px;
}
.fk-plot-window-control__label {
    padding-right: 4px;
}
.fk-plot-window-control__label-padding {
    padding-left: 6px;
}
.fk-plot-window-numeric-control {
    display: flex;
    flex-direction: row;
    align-items: baseline;
    flex: 0 1 auto;
    padding-left: 6px;
    padding-right: 6px;
    width: 140px;
}
.fk-weavess-wrapper-wrapper {
    flex: 0 1 auto;
    position: relative;
    height: 330px;
}
.fk-channel-list {
    padding: 1rem;
    overflow-y: auto;
    max-height: 200px;
}
.fk-plots-wrapper-1 {
    width: 100%;
    height: 100%;
    color: $gray4;
}
.fk-plots-wrapper-2 {
    height: 100%;
    width: 100%;
    display: flex;
    min-width: fit-content;
    min-height: 332px;
    flex-direction: column;
}
.fk-image-and-details-container {
    padding: 0.5rem;
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: row;
    user-select: none;
    flex-wrap: wrap;
}
.fk-color-scale {
    position: relative;
    width: 100%;
    height: 100%;
}
.fk-color-scale__xaxis {
    position: absolute;
    right: 0px;
    left: 0px;
    bottom: 0px;
    height: 35px;
    top: 18px
}
.fk-color-scale__canvas {
    position: absolute;
    right: 25px;
    left: 25px;
    top: 0px;
    z-index: 0;
}
.fk-color-scale__units {
    margin-top: 2px;
    position: absolute;
    right: 0px;
    font-size: 11px;
}
.fk-color-scale__button:hover {
    padding: 1px 4px 1px 4px;
    cursor: pointer;
    background-color: white;
    color: black;
    border-radius: 4px;
}

.fk-units-picker {
    display: flex;
}
.fk-units-picker__label {
    margin-right: 8px;
}
.fk-units-picker__radio-group {
    display: flex;
}

.continous-fk-popover {
    padding: 8px 12px 8px 12px;
}

.continous-fk-popover-panel {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
}

.popover-section {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    width: 400px;
    margin-bottom: 8px;

}
.popover-header {
    display: flex;
    justify-content: space-between;
}
.popover-header__text {
    font-size: 14px;
    font-weight: bold;
    line-height: 28px;
}

.popover-body-text {
    font-size: 14px;
    color: rgb(167, 182, 194);
}

.popover-button-group {
    display: flex;
    justify-content: center;
    margin-bottom: 8px;
    margin-top: 4px;
}
.continous-fk-popover-apply-cancel-row {
    display: flex;
    width: 100%;
    justify-content: flex-end;
}

.continous-fk-popover-apply {
    margin-right: 8px;
}

.popover-header__widget {
    display: flex;
    align-items: center;
}

.fk-frequency-thumbnails{
    display: flex;
    flex-wrap: wrap;
    width: 229px;
    padding: .44rem;
}

.fk-properties__frequency-low-high-inputs {
    width: 100px;
}