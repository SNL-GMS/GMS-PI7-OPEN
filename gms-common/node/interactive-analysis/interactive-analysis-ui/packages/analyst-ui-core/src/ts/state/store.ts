import * as Redux from 'redux';
import { createLogger } from 'redux-logger';
import { default as thunk } from 'redux-thunk';
import { getElectron, getElectronEnhancer } from '../util/electron-util';
import { IS_NODE_ENV_DEVELOPMENT } from '../util/environment';
import { isWindowDefined } from '../util/window-util';
import { initialAppState } from './initial-state';
import { Reducer } from './root-reducer';
import { AppState } from './types';

declare var window;

const windowIsDefined = isWindowDefined();

const windowRedux: Window & {
  __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?(a: any): void;
} = windowIsDefined ? window : undefined;

const electron = getElectron();
const electronEnhancer = getElectronEnhancer();

const composeEnhancers = (windowRedux) ?
  (windowRedux as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ != null
    ? (windowRedux as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    : Redux.compose
  : Redux.compose;

const configureStore = (initialState?: Partial<AppState> | void) => {
  let store: Redux.Store;

  const middlewares = [];
  middlewares.push(thunk);
  if (IS_NODE_ENV_DEVELOPMENT) {
    const logger = createLogger({
      collapsed: true,
      duration: true,
      timestamp: false,
      level: 'info',
      logger: console,
      logErrors: true,
      diff: false
    });
    middlewares.push(logger);
  }

  const enhancers =
    (electron && electron !== undefined && electronEnhancer) ?
      composeEnhancers(
        Redux.applyMiddleware(...middlewares),
        // Must be placed after any enhancers which dispatch
        // their own actions such as redux-thunk or redux-saga
        electronEnhancer({
          dispatchProxy: a => store.dispatch(a),
        }))
      :
      composeEnhancers(Redux.applyMiddleware(...middlewares));

  store = Redux.createStore(
    Reducer,
    initialState as any,
    enhancers);
  return store;
};

// tslint:disable-next-line:no-default-export
export const createStore = (): Redux.Store<AppState> => {
  const store = configureStore(initialAppState);
  return store;
};
