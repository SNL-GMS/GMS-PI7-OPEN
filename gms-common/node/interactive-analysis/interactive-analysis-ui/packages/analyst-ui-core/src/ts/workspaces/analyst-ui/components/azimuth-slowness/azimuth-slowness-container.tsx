
import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import * as ReactRedux from 'react-redux';
import { bindActionCreators } from 'redux';
import { systemConfig, userPreferences } from '~analyst-ui/config';
import {
  EventQueries,
  EventTypes,
  FkMutations,
  SignalDetectionQueries,
  SignalDetectionTypes,
  StationQueries,
  StationTypes,
  WaveformQueries
} from '~graphql/';
import { AnalystWorkspaceActions } from '~state/';
import { AppState } from '~state/types';
import { AzimuthSlowness } from './azimuth-slowness-component';
import { AzimuthSlownessProps, AzimuthSlownessReduxProps } from './types';

/**
 * Container component for Azimuth Slowness
 * Handles mapping of state/props through redux/apollo
 */

/**
 *  Mapping between the current redux state and props for the Azimuth Slowness Display
 */
const mapStateToProps = (state: AppState): Partial<AzimuthSlownessReduxProps> => ({
  currentStageInterval: state.analystWorkspaceState.currentStageInterval,
  currentTimeInterval: (state.analystWorkspaceState.currentStageInterval) ?
    state.analystWorkspaceState.currentStageInterval.interval.timeInterval : undefined,
  analystActivity: (state.analystWorkspaceState.currentStageInterval) ?
    state.analystWorkspaceState.currentStageInterval.interval.activityInterval.analystActivity : undefined,
  selectedSdIds: state.analystWorkspaceState.selectedSdIds,
  openEventId: state.analystWorkspaceState.openEventId,
  selectedSortType: state.analystWorkspaceState.selectedSortType,
  sdIdsToShowFk: state.analystWorkspaceState.sdIdsToShowFk,
  measurementMode: state.analystWorkspaceState.measurementMode,
  channelFilters: state.analystWorkspaceState.channelFilters,
  client: state.apolloClient.client
});

/**
 * Map actions dispatch callbacks into this component as props
 */
const mapDispatchToProps = (dispatch): Partial<AzimuthSlownessReduxProps> =>
  bindActionCreators(
    {
      setSelectedSdIds: AnalystWorkspaceActions.setSelectedSdIds,
      setSdIdsToShowFk: AnalystWorkspaceActions.setSdIdsToShowFk
    } as any,
    dispatch
  );

/**
 * The higher-order component react-redux(react-apollo(AzimuthSlowness))
 */
export const ReduxApolloAzimuthSlowness: React.ComponentClass<Pick<{}, never>> = compose(
  ReactRedux.connect(mapStateToProps, mapDispatchToProps),
  graphql(StationQueries.defaultStationsQuery, { name: 'defaultStationsQuery' }),
  graphql(WaveformQueries.defaultWaveformFiltersQuery, { name: 'defaultWaveformFiltersQuery' }),
  graphql(SignalDetectionQueries.signalDetectionsByStationQuery, {
    options: (props: AzimuthSlownessProps) => {
      // Get signal detections in the current interval
      const variables: SignalDetectionTypes.SignalDetectionsByStationQueryArgs = {
        stationIds: props.defaultStationsQuery.defaultStations.map(station => station.id),
        timeRange: {
          startTime: Number(props.currentTimeInterval.startTimeSecs) -
            systemConfig.additionalDataToLoadOnInitialLoad(props.analystActivity),
          endTime: Number(props.currentTimeInterval.endTimeSecs) +
            systemConfig.additionalDataToLoadOnInitialLoad(props.analystActivity),
        }
      };
      return {
        variables,
      };
    },
    skip: (props: AzimuthSlownessProps) => !props.currentTimeInterval ||
    !props.defaultStationsQuery.defaultStations,
    name: 'signalDetectionsByStationQuery'
  }),
  graphql(StationQueries.distanceToSourceForDefaultStationsQuery, {
    options: (props: AzimuthSlownessProps) => {
      const variables: StationTypes.DistanceToSourceForDefaultStationsQueryArgs = {
        distanceToSourceInput: {
          sourceId: props.openEventId,
          sourceType: 'Event',
          distanceUnits: userPreferences.distanceUnits
        }
      };
      return {
        variables
      };
    },
    skip: (props: AzimuthSlownessProps) => !props.currentTimeInterval || !props.openEventId,
    name: 'distanceToSourceForDefaultStationsQuery'
  }),
  graphql(EventQueries.eventsInTimeRangeQuery, {
    options: (props: AzimuthSlownessProps) => {
      const variables: EventTypes.EventsInTimeRangeQueryArgs = {
        timeRange: {
          startTime: Number(props.currentTimeInterval.startTimeSecs) -
            systemConfig.additionalEventDataToLoadInitially(props.analystActivity),
          endTime: Number(props.currentTimeInterval.endTimeSecs) +
            systemConfig.additionalDataToLoadOnInitialLoad(props.analystActivity)
        }
      };
      return {
        variables,
      };
    },
    skip: (props: AzimuthSlownessProps) => !props.currentTimeInterval,
    name: 'eventsInTimeRangeQuery'
  }),

  graphql(FkMutations.computeFksMutation, { name: 'computeFks' }),
  graphql(FkMutations.setFkWindowLeadMutation, { name: 'setWindowLead' }),
  graphql(FkMutations.markFksReviewedMutation, { name: 'markFksReviewed' })

)(AzimuthSlowness);
