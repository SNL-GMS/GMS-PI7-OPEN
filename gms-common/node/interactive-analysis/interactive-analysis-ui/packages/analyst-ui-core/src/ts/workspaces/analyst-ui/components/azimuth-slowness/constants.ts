import { FkTypes } from '~graphql/';
import { LeadLagPairAndString } from './types';

// Width of fk properties, used to calculate size allocation for fk rendering
export const MAX_WIDTH_OF_FK_PROPERTIES_PX = 493;

// Height of fk plots, used to calcualte size allocation for fk rendering
export const MAX_HEIGHT_OF_FK_PLOTS_PX = 424;

// Width/height of y and x axis respectively
export const SIZE_OF_FK_RENDERING_AXIS_PX = 35;

// The height of the rendering needs to be 8 pixels smaller than the width
export const FK_RENDERING_HEIGHT_OFFSET = 8;

/** 
 * Hard-coded lead/lag options
 */

export const LeadLagPairsValues: FkTypes.WindowParameters[] = [
    {
      leadSeconds: 1,
      lengthSeconds: 4,
      stepSize: undefined
    },
    {
      leadSeconds: 1,
      lengthSeconds: 6,
      stepSize: undefined
    },
    {
      leadSeconds: 1,
      lengthSeconds: 9,
      stepSize: undefined
    },
    {
      leadSeconds: 1,
      lengthSeconds: 11,
      stepSize: undefined

    }
  ];
export enum LeadLagPairs {
    LEAD_1_DURATION_4 = 'Lead: 1, Dur: 4',
    LEAD_1_DURATION_6 = 'Lead: 1, Dur: 6',
    LEAD_1_DURATION_9 = 'Lead: 1, Dur: 9',
    LEAD_1_DURATION_11 = 'Lead: 1, Dur: 11',
  }
export enum LeadLagPairsAndCustom {
  LEAD_1_DURATION_4 = 'Lead: 1, Dur: 4',
  LEAD_1_DURATION_6 = 'Lead: 1, Dur: 6',
  LEAD_1_DURATION_9 = 'Lead: 1, Dur: 9',
  LEAD_1_DURATION_11 = 'Lead: 1, Dur: 11',
  CUSTOM = 'Custom'
}
export const LeadLagValuesAndDisplayString: LeadLagPairAndString[] = [{
  leadLagPairs: LeadLagPairs.LEAD_1_DURATION_4,
  windowParams: LeadLagPairsValues[0]
},
{
  leadLagPairs: LeadLagPairs.LEAD_1_DURATION_6,
  windowParams: LeadLagPairsValues[1]
},
{
  leadLagPairs: LeadLagPairs.LEAD_1_DURATION_9,
  windowParams: LeadLagPairsValues[2]
},
{
  leadLagPairs: LeadLagPairs.LEAD_1_DURATION_11,
  windowParams: LeadLagPairsValues[3]
}];
