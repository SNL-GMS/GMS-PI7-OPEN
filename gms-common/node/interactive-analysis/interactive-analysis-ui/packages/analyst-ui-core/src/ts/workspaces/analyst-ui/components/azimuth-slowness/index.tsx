export { AzimuthSlowness } from './azimuth-slowness-component';
