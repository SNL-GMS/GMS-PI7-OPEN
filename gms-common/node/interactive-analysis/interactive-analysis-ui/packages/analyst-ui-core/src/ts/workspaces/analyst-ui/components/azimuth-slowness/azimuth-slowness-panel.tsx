import { ContextMenu } from '@blueprintjs/core';
import * as React from 'react';
import { getFkData } from '~analyst-ui/common/utils/fk-utils';
import { FkTypes } from '~graphql';
import { FkConfiguration } from '~graphql/fk/types';
import { UILogger } from '~util/log/logger';
import { FkThumbnailBlueprintContextMenu } from './components/context-menus/fk-context-menu';
import { FkDisplay } from './components/fk-display/fk-display';
import { FkThumbnailList } from './components/fk-thumbnail-list/fk-thumbnail-list';
import { FkThumbnailsControls } from './components/fk-thumbnail-list/fk-thumbnails-controls';
import { createFkInput, fkNeedsReview, getSortedSignalDetections } from './components/fk-util';
import { AzimuthSlownessPanelProps, FkParams } from './types';

/**
 * An intermediary between AzimuthSlownessComponent and the other components so that event handling is simpler 
 */
export class AzimuthSlownessPanel extends React.Component<AzimuthSlownessPanelProps, {}> {
  /** Used to constrain the max width of the thumbnail drag resize */
  private azimuthSlownessContainer: HTMLDivElement;

  /** Used to drag & resize this element */
  private fkThumbnailsContainer: HTMLDivElement;

  /** The inner container for the thumbnail */
  private fkThumbnailsInnerContainer: HTMLDivElement;

  /**
   * Invoked when the component mounted.
   */
  public componentDidMount() {
        this.props.adjustFkInnerContainerWidth(this.fkThumbnailsContainer, this.fkThumbnailsInnerContainer);
    }

  public componentDidUpdate(prevProps: AzimuthSlownessPanelProps) {
      if (this.props.fkThumbnailColumnSizePx !== prevProps.fkThumbnailColumnSizePx) {
        this.props.adjustFkInnerContainerWidth(this.fkThumbnailsContainer, this.fkThumbnailsInnerContainer);
      }
    }

  public render () {
      const anyDislayedFksNeedReview = this.getReviewableSds().length > 0;
      return (
        <div
          ref={ref => this.azimuthSlownessContainer = ref}
          className="azimuth-slowness-container"
          tabIndex={0}
          onKeyDown={this.onKeyDown}
        >
        <div
          ref={ref => this.fkThumbnailsContainer = ref}
          className="azimuth-slowness-thumbnails"
          style={{ width: `${this.props.fkThumbnailColumnSizePx}px` }}
        >
          <div
            className="azimuth-slowness-thumbnails__control-container"
          >
            <FkThumbnailsControls
              updateFkThumbnail={this.props.updateFkThumbnailSize}
              updateFkFilter={this.props.updateFkFilter}
              anyDislayedFksNeedReview={anyDislayedFksNeedReview}
              onlyOneFkIsSelected={this.props.selectedSdIds.length === 1}
              widthPx={this.props.fkThumbnailColumnSizePx}
              nextFk={() => {this.nextFk(); }}
              currentFilter={this.props.filterType}
              clearSelectedUnassociatedFks={this.clearSelectedUnassociatedFks}
            />
          </div>
          <div
            className="azimuth-slowness-thumbnails__wrapper-1"
          >
            <div
              ref={ref => this.fkThumbnailsInnerContainer = ref}
              className="azimuth-slowness-thumbnails__wrapper-2"
            >
              <FkThumbnailList
                thumbnailSizePx={this.props.fkThumbnailSizePx}
                sortedSignalDetections={this.props.signalDetectionsToDraw}
                unassociatedSdIds={
                  this.props.signalDetectionsToDraw
                  .filter(sdToDraw => !this.props.associatedSignalDetections.find(aSd => aSd.id === sdToDraw.id))
                  .map(sd => sd.id)
                }
                signalDetectionIdsToFeaturePrediction={this.props.signalDetectionsIdToFeaturePredictions}
                distanceToSource={this.props.distanceToSource}
                selectedSdIds={this.props.selectedSdIds}
                setSelectedSdIds={this.props.setSelectedSdIds}
                selectedSortType={this.props.selectedSortType}
                clearSelectedUnassociatedFks={this.clearSelectedUnassociatedFks}
                fkUnitsForEachSdId={this.props.fkUnitsForEachSdId}
                markFksForSdIdsAsReviewed={this.markFksForSdIdsAsReviewedIfTheyNeedToBeAccepted}
                showFkThumbnailContextMenu={this.showFkThumbnailMenu}
              />
            </div>
          </div>
        </div>
          {/* drag handle divider */}
          <div
            className="azimuth-slowness-divider"
            onMouseDown={this.onThumbnailDividerDrag}
          >
            <div
              className="azimuth-slowness-divider__spacer"
            />
          </div>
          <FkDisplay
            measurementMode={this.props.measurementMode}
            defaultStations={this.props.defaultStations}
            defaultWaveformFilters={this.props.defaultWaveformFilters}
            channelFilters={this.props.channelFilters}
            eventsInTimeRange={this.props.eventsInTimeRange}
            currentOpenEvent={this.props.openEvent}
            signalDetection={this.props.displayedSignalDetection}
            signalDetectionsByStation={this.props.signalDetectionsByStation}
            signalDetectionFeaturePredictions={this.props.featurePredictionsForDisplayedSignalDetection}
            widthPx={this.props.fkDisplayWidthPx}
            numberOfOutstandingComputeFkMutations={this.props.numberOfOutstandingComputeFkMutations}
            heightPx={this.props.fkDisplayHeightPx}
            multipleSelected={this.props.selectedSdIds && this.props.selectedSdIds.length > 1}
            anySelecteced={this.props.selectedSdIds && this.props.selectedSdIds.length > 0}
            userInputFkFrequency={this.props.userInputFkFrequency}
            fkUnit={this.props.fkUnitForDisplayedSignalDetection}
            userInputFkWindowParameters={this.props.userInputFkWindowParameters}
            onNewFkParams={this.onNewFkParams}
            changeUserInputFks={this.props.changeUserInputFks}
            setFkUnitForSdId={this.props.setFkUnitForSdId}
            onSetWindowLead={this.props.setWindowLead}
            fkFrequencyThumbnails={this.props.fkFrequencyThumbnails}
          />
        </div>
      );
    }

  /**
   * Selects the next fk that needs review
   */
  private readonly nextFk = () => {
    if (!this.props.displayedSignalDetection) {
      throw Error('Selected Signal Detection not found in client cache');
    }
    const reviewableSds = this.getReviewableSds();
    if (reviewableSds.length === 1 &&
      reviewableSds.find(sd => sd.id === this.props.displayedSignalDetection.id)) {
      this.props.setSelectedSdIds([]);
      this.markFksForSdIdsAsReviewedIfTheyNeedToBeAccepted([this.props.displayedSignalDetection.id]);

    } else {
      const needsReviewSds = reviewableSds
                      .filter(sd => sd.id !== this.props.displayedSignalDetection.id);
      const sortedNeedsReviewSds =
        getSortedSignalDetections(needsReviewSds, this.props.selectedSortType, this.props.distanceToSource);
      this.props.setSelectedSdIds(
        [sortedNeedsReviewSds[0].id]
      );
      this.markFksForSdIdsAsReviewedIfTheyNeedToBeAccepted([this.props.displayedSignalDetection.id]);
    }
  }

  /**
   * Handles keypresses on az slow
   * @param e keyboard event
   */
  private readonly onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
      if (e.nativeEvent.code === 'KeyN' && (e.metaKey || e.ctrlKey)) {
        if (this.getReviewableSds().length > 0 && this.props.selectedSdIds.length === 1) {
            this.nextFk();
        }
      }
    }

  /**
   * Clear selected UnassociatedFks
   */
  private readonly clearSelectedUnassociatedFks = (): void => {
    const selectedAndUnnasociatedSdIds = this.props.selectedSdIds.filter(sdId =>
      !this.props.associatedSignalDetections.find(assocSd => assocSd.id === sdId)
    );
    const selectedAndAsociatedSdIds = this.props.selectedSdIds.filter(sdId =>
      this.props.associatedSignalDetections.find(assocSd => assocSd.id === sdId)
    );
    const sdsIdsToShowFk =
      this.props.sdIdsToShowFk.filter(sdId => !selectedAndUnnasociatedSdIds.find(unSdId => unSdId === sdId));
    this.props.setSelectedSdIds(selectedAndAsociatedSdIds);
    this.props.setSdIdsToShowFk(sdsIdsToShowFk);
  }

  /**
   * Shows the fk thumbnail menu
   * 
   * @param x offset from left
   * @param y offset from top
   */
  private readonly showFkThumbnailMenu = (x: number, y: number) => {
      const selectedAndUnnasociatedSdIds = this.props.selectedSdIds.filter(sdId =>
        !this.props.associatedSignalDetections.find(assocSd => assocSd.id === sdId)
      );
      const stageIntervalContextMenu = FkThumbnailBlueprintContextMenu(
      this.clearSelectedUnassociatedFks,
      selectedAndUnnasociatedSdIds.length > 0);
      ContextMenu.show(
      stageIntervalContextMenu, {
        left: x,
        top: y
      });
    }

  /**
   * Start a drag on mouse down on the divider
   */
  private readonly onThumbnailDividerDrag = (e: React.MouseEvent<HTMLDivElement>) => {
    let prevPosition = e.clientX;
    let currentPos = e.clientX;
    let diff = 0;
    const maxWidthPct = 0.8;
    const maxWidthPx = this.azimuthSlownessContainer.clientWidth * maxWidthPct;

    const onMouseMove = (e2: MouseEvent) => {
      currentPos = e2.clientX;
      diff = currentPos - prevPosition;
      prevPosition = currentPos;
      const widthPx = this.fkThumbnailsContainer.clientWidth + diff;
      if (widthPx < maxWidthPx) {
          this.props.setFkThumbnailColumnSizePx(widthPx);
      }
    };

    const onMouseUp = (e2: MouseEvent) => {
      document.body.removeEventListener('mousemove', onMouseMove);
      document.body.removeEventListener('mouseup', onMouseUp);
    };

    document.body.addEventListener('mousemove', onMouseMove);
    document.body.addEventListener('mouseup', onMouseUp);
    }

  /**
   * Handles new FK Request when frequency and/or window params change
   */
  private readonly onNewFkParams = async (
    sdId: string, fkParams: FkParams, fkConfiguration: FkConfiguration): Promise<void> => {
    const fkData = getFkData(this.props.displayedSignalDetection.currentHypothesis.featureMeasurements);
    if (fkData) {
      const fkInput: FkTypes.FkInput = createFkInput(
        this.props.displayedSignalDetection,
        fkParams.frequencyPair,
        fkParams.windowParams,
        fkConfiguration
      );
      this.props.computeFkAndUpdateState(fkInput)
      .catch(error => UILogger.error(`Failed onNewFkParams: ${error}`));
    }
  }

  /**
   * @param sdIds a list of candidate sd ids
   */
  private readonly markFksForSdIdsAsReviewedIfTheyNeedToBeAccepted
    = (sdIds: string[]) => {

      const sdIdsThatCanBeReviewed = this.getReviewableSds()
        .map(sd => sd.id);
      const sdIdsToReview =
        sdIds.filter(sdId => sdIdsThatCanBeReviewed.find(sdReviewableId => sdReviewableId === sdId));
      if (sdIdsToReview.length > 0) {
        this.props.markFksForSdIdsAsReviewed(sdIdsToReview);
      }
  }
  /**
   * Gets the list of fks that need review and are associated to the currently open event
   */
  private readonly getReviewableSds = () => this.props.signalDetectionsToDraw.filter(
    sdToDraw =>
      this.props.associatedSignalDetections.find(aSd => aSd.id === sdToDraw.id) &&
      fkNeedsReview(sdToDraw)
    )
}
