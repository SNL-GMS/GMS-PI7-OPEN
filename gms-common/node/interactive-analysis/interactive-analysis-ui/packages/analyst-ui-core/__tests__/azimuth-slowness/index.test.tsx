
import * as React from 'react';
import { ApolloProvider } from 'react-apollo';
import { Provider } from 'react-redux';
// tslint:disable-next-line:max-line-length
import { ReduxApolloAzimuthSlowness } from '~analyst-ui/components/azimuth-slowness/azimuth-slowness-container';
import { createStore } from '~state/store';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');
// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Adapter = require('enzyme-adapter-react-16');

jest.mock('graphql'); // graphql is now a mock constructor

const store =  createStore();

it('should render a ReduxApolloAzimuthSlowness component correctly', () => {
  Enzyme.configure({ adapter: new Adapter() });
  const wrapper = Enzyme.shallow(
    <ApolloProvider client={store.getState().apolloClient.client}>
      <Provider store={store}>
        <ReduxApolloAzimuthSlowness />
      </Provider>
    </ApolloProvider>
  );
  wrapper.dive();
  expect(wrapper)
    .toMatchSnapshot();
});
