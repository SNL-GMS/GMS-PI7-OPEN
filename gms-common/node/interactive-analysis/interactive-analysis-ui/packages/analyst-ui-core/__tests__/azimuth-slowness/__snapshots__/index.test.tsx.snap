// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a ReduxApolloAzimuthSlowness component correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Apollo(Apollo(Apollo(Apollo(Apollo(Apollo(Apollo(Apollo(AzimuthSlowness))))))))) />
</Provider>
`;
